File: 2015/01/main.js
export function part1(data) {
    let floor = 0;

    for (let i = 0; i < data.length; i++) {
        const element = data[i];
        if (element === '(') {
            floor++;
        } else {
            floor--;
        }
    }

    return floor;
}

export function part2(data) {
    let floor = 0;

    for (let i = 0; i < data.length; i++) {
        const element = data[i];
        if (element === '(') {
            ++floor;
        } else {
            --floor;
        }
        if (floor < 0) {
            return i + 1;
        }
    }
}

File: 2015/01/main.go
package main

import (
	"fmt"

	"2015/utils"
)

func main() {
	data, err := utils.Read("input.txt")
	if err != nil {
		fmt.Println("Error reading input file:", err)
		return
	}

	fmt.Println(part1(data))
	fmt.Println(part2(data))
}

func part1(data string) int {
	floor := 0

	for _, c := range data {
		if c == '(' {
			floor++
		} else {
			floor--
		}
	}

	return floor
}

func part2(data string) int {
	floor := 0

	for i, c := range data {
		if c == '(' {
			floor++
		} else {
			floor--
		}
		if floor == -1 {
			return i + 1
		}
	}

	return floor
}
